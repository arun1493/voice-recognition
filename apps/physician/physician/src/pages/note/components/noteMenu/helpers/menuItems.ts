import i18n from "i18next";

import { MenuItem } from "../types";

const getMenuItems = ({ isDeleteDisabled, onDelete, onPrint }: any): MenuItem[] => {
  return [
    {
      icon: "print",
      label: i18n.t("components.noteMenu.options.print"),
      onClick: onPrint
    },
    {
      disabled: isDeleteDisabled,
      icon: "trash",
      label: i18n.t("components.noteMenu.options.delete"),
      onClick: onDelete
    }
  ];
};

export { getMenuItems };
