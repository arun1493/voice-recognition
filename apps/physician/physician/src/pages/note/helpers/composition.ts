import _debounce from "lodash/debounce";
import { ClientEmitEventTypes } from "@suki-web/suki-business";

import { NoteSection } from "../../../types";
import { sectionReader } from "../../../readers";
import { USER_REQUEST_TYPE } from "../../../constants";

const saveComposition = _debounce(
  (sectionData: NoteSection, compositionId: string, noteTypeId: string, transmitEvent: any) => {
    const sectionId = sectionReader.sectionId(sectionData);

    transmitEvent({
      type: ClientEmitEventTypes.USER_REQUEST,
      data: { compositionId, sectionData, sectionId, noteTypeId, requestType: USER_REQUEST_TYPE.UPDATE_SECTION }
    });
  },
  800
);

export { saveComposition };
