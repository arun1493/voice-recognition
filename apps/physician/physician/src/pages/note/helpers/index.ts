export * from "./problem";
export * from "./scripts";
export * from "./sections";
export * from "./composition";
export * from "./scheduleOrList";
export * from "./problemSection";
export * from "./underscoreOccurances";
