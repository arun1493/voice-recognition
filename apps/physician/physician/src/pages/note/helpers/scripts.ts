import _findIndex from "lodash/findIndex";

import { NoteSection, Script, ScriptBlock } from "../../../types";
import { INDEX_NOT_FOUND, SUKI_ORGANIZATION_ID } from "../../../constants";

const renderScriptContentBlock = (section: NoteSection) => (block: ScriptBlock) => {
  return block?.sectionId === section?.id || block?.sectionId === SUKI_ORGANIZATION_ID;
};

const checkIfSectionHasScripts = (section: NoteSection) => (scriptContent: Script) => {
  const { blocks } = scriptContent;
  const sectionIndex = _findIndex(blocks, renderScriptContentBlock(section));
  return sectionIndex > INDEX_NOT_FOUND;
};

const createModules = (
  scripts: Script[],
  sectionName: string,
  activeSectionId: string,
  clickInsertScript: (scriptId: string, activeSectionId: string) => void
) => {
  return {
    history: {
      userOnly: false,
      maxStack: 100,
      delay: 1000
    },
    toolbar: false,
    keyboard: {
      bindings: {
        // Tab was replaced by key stroke('\t') in editor
        // Return true to let later handlers be called
        tab: {
          key: 9,
          handler: () => true
        }
      }
    },
    mention: {
      allowedChars: /^[A-Za-z\sÅÄÖåäö]*$/,
      mentionDenotationChars: ["+"],
      dataAttributes: ["cy"],
      positioningStrategy: "fixed",
      source: (searchTerm: string, renderList: any) => {
        if (!scripts) return;

        const scriptValues =
          scripts &&
          scripts.map((s, i) => {
            const secName = sectionName || "new";
            const scriptName = s.tags[0].name.value;
            return {
              cy: `section-${secName.replace(/\s/g, "_")}-script-${scriptName.replace(/\s/g, "_")}`,
              id: s.id,
              value: scriptName
            };
          });
        const values = scriptValues;
        if (searchTerm.length === 0) {
          renderList(values, searchTerm);
        } else {
          const matches = [];
          for (let i = 0; i < values.length; i++) {
            if (~values[i].value.toLowerCase().indexOf(searchTerm.toLowerCase())) {
              matches.push(values[i]);
            }
            renderList(matches, searchTerm);
          }
        }
      },
      onSelect: async ({ id }: any) => {
        clickInsertScript(id, activeSectionId);
      }
    }
  };
};

export { checkIfSectionHasScripts, createModules, renderScriptContentBlock };
