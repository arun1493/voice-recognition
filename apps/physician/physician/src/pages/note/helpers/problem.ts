import _get from "lodash/get";
import _isEmpty from "lodash/isEmpty";
import _cloneDeep from "lodash/cloneDeep";
import { EMPTY_ARRAY, SectionUpdateTypes } from "@suki-web/suki-business";

import {
  appointmentReader,
  compositionReader,
  diagnosisReader,
  metadataReader,
  patientReader,
  sectionReader
} from "../../../readers";
import { searchDiagnoses } from "../../../api";
import { getSectionIcdCode } from "../../../helpers";
import { PROBLEM_STATUSES } from "../../../constants";
import { Composition, NoteSection } from "../../../types";

const getExistingDiagnoses = (sections: NoteSection[]) => {
  const sectionsWithICDCode = sections.filter(getSectionIcdCode);
  return sectionsWithICDCode.map((section: NoteSection) => {
    const diagnosis = sectionReader.diagnosis(section);
    return {
      snomedDescription: sectionReader.sectionName(section),
      icdCode: diagnosisReader.icdCode(diagnosis),
      icdDescription: diagnosisReader.icdDescription(diagnosis),
      snomedCode: diagnosisReader.snomedCode(diagnosis)
    };
  });
};

const getDiagnosisNames = (sections: NoteSection[]) => {
  const problemSections = sections.filter(sectionReader.diagnosis);

  return problemSections.map(sectionReader.sectionName);
};

const getAppointmentDetails = (composition: Composition) => {
  const metadata = compositionReader.metadata(composition);
  const appointment = metadataReader.appointment(metadata);
  const patient = metadataReader.patient(metadata);

  return {
    emrPatientId: patientReader.mrn(patient),
    emrAppointmentDepartmentId: appointmentReader.emrDepartmentId(appointment),
    emrEncounterId: appointmentReader.emrEncounterId(appointment),
    sukiAppointmentId: appointmentReader.id(appointment),
    sukiPatientId: patientReader.id(patient)
  };
};

const autoPopulateIcdCodes = async (section: NoteSection) => {
  // Auto Populate ICD Codes
  const sectionName = sectionReader.sectionName(section);

  if (_isEmpty(sectionName)) return;

  const searchResults = await searchDiagnoses(sectionName);
  const diagnoses = _get(searchResults, "data.diagnoses", EMPTY_ARRAY); //.0.diagnosis

  // diagnoses should not have any children
  if (
    _isEmpty(diagnoses) ||
    !_isEmpty(_get(diagnoses, "0.children")) ||
    _isEmpty(_get(diagnoses, "0.diagnosis")) ||
    diagnoses?.length > 1
  )
    return;

  // we have exact match now
  const suggestedDiagnosis = _get(diagnoses, "0.diagnosis");
  const updatedSection = _cloneDeep(section);
  updatedSection.diagnosisEntry = {
    diagnosis: {
      ...suggestedDiagnosis,
      problemLabel: sectionName
    },
    confirmStatus: PROBLEM_STATUSES.UNCONFIRMED
  };
  updatedSection.updateType = SectionUpdateTypes.ONLY_DIAGNOSIS;

  return updatedSection;
};

// This methods returns rows and cols span for the the textarea in problem name to adjust its size to accommodate large names
// [TODO][Uttam]: Fix the logic for height adjustment for textarea
const getTextAreaRowsAndColsCount = (name: string, element: HTMLTextAreaElement) => {
  const elementWidth = element?.clientWidth || 300;

  let divider = 10;
  if (elementWidth <= 170) divider = 16;
  else if (elementWidth <= 185) divider = 15;
  else if (elementWidth <= 210) divider = 13;
  else if (elementWidth <= 280) divider = 12;
  else if (elementWidth <= 305) divider = 11;

  const colsCount = Math.ceil(elementWidth / divider);
  const rowsCount = Math.ceil(name.length / colsCount);

  return { rowsCount, colsCount };
};

export {
  autoPopulateIcdCodes,
  getAppointmentDetails,
  getDiagnosisNames,
  getExistingDiagnoses,
  getTextAreaRowsAndColsCount
};
