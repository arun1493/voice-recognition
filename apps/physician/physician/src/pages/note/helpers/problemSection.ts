import _map from "lodash/map";

import { sectionReader } from "../../../readers";
import { getProblemSections } from "../../../helpers";
import { Composition, NoteSection } from "../../../types";

const getSectionIdNameString = (section: NoteSection) =>
  `${sectionReader.sectionId(section)}:${sectionReader.sectionName(section)}`;

const getProblemSectionsIdNameMap = (composition: Composition) => {
  const problemSections = getProblemSections(composition);
  return _map(problemSections, getSectionIdNameString);
};

export { getProblemSectionsIdNameMap };
