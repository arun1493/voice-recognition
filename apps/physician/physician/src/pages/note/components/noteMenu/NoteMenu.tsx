import _map from "lodash/map";
import { FunctionComponent } from "react";
import { useParams } from "react-router-dom";
import { useTranslation } from "react-i18next";
import { useDispatch, useSelector } from "react-redux";
import { DropDownMenu, IconType } from "@suki-web/suki-ui";

import { MenuItem } from "./types";
import { getMenuItems } from "./helpers";
import { TOAST_TYPES } from "../../../../constants";
import { toastsActions } from "../../../../redux/toast";
import { modalsActions } from "../../../../redux/modals";
import { StyledNoteMenuContainer } from "./noteMenu.styled";
import { isNoteCompleted, printNote } from "../../../../helpers";
import { selectCurrentComposition } from "../../../../redux/note";
import { NoteMenuListItem } from "../../../../components/molecules";
import { compositionReader, metadataReader } from "../../../../readers";

interface NoteMenuProps {
  isReadOnlyMode: boolean;
  onClose: () => void;
}

type OptionsItem = {
  icon: IconType;
  isDisabled: boolean;
  label: string;
  onClick?: (e: { preventDefault: () => void }) => void;
  path?: string;
};

const NoteMenu: FunctionComponent<NoteMenuProps> = ({ isReadOnlyMode, onClose }) => {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const { noteId } = useParams<{ noteId: string }>();
  const compositionData = useSelector(selectCurrentComposition);
  const compositionId = compositionReader.compositionId(compositionData);
  const metadata = compositionReader.metadata(compositionData);
  const noteStatus = metadataReader.status(metadata);
  const isPartialDictationComposition = metadataReader.partialDictationComposition(metadata);

  const noteParams = isNoteCompleted(noteStatus) ? { noteId } : { compositionId };

  const handlePrintMenuItemClick = () => {
    printNote(noteParams);
    onClose();
  };

  const handleDeleteMenuItemClick = () => {
    onClose();

    if (isPartialDictationComposition) {
      dispatch(
        toastsActions.add({
          message: t("modals.deleteModal.deletePartialDictationCompositionError"),
          type: TOAST_TYPES.ERROR
        })
      );
      return;
    }

    if (!isReadOnlyMode) {
      dispatch(modalsActions.showDeleteNoteModal());
    }
  };

  const menuItems = getMenuItems({
    isDeleteDisabled: isReadOnlyMode,
    onPrint: handlePrintMenuItemClick,
    onDelete: handleDeleteMenuItemClick
  });

  const renderMenuItem = ({ disabled, icon, label, onClick }: MenuItem, id: number) => {
    const key = `note-menu-item-${id}`;
    return (
      <NoteMenuListItem data-cy={key} disabled={disabled} icon={icon} key={key} onClick={onClick} testId={key}>
        {label}
      </NoteMenuListItem>
    );
  };

  return (
    <DropDownMenu>
      <StyledNoteMenuContainer data-cy="note-menu">{_map(menuItems, renderMenuItem)}</StyledNoteMenuContainer>
    </DropDownMenu>
  );
};

export { NoteMenu };

export default NoteMenu;
