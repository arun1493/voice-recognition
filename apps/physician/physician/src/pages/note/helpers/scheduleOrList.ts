import store from "store2";
import _get from "lodash/get";
import _find from "lodash/find";
import _isEmpty from "lodash/isEmpty";
import { differenceInCalendarDays } from "date-fns";
import { AppointmentTypes } from "@suki-web/suki-business";

import { isEpicEMR } from "../../../helpers";
import { LOCAL_STORE } from "../../../constants";
import { noteActions } from "../../../redux/note";
import { modalsActions } from "../../../redux/modals";
import { Dispatch, Metadata, Patients } from "../../../types";
import { getAppointments, getPatientList } from "../../../api";
import { appointmentReader, metadataReader, patientReader } from "../../../readers";

async function handleAddPatientToScheduleOrList(
  metadata: Metadata,
  dispatch: Dispatch,
  isEmr: boolean,
  isNewNote: boolean
): Promise<void> {
  const appointment = metadataReader.appointment(metadata);
  const patient = metadataReader.patient(metadata);
  const emrType = metadataReader.emrType(metadata) || store.get(LOCAL_STORE.EMR_TYPE);
  const patientId = patientReader.id(patient);

  if (_isEmpty(patientId)) {
    return;
  }

  // if its not a new note don't show the dialog
  if (!isNewNote) {
    return;
  }

  if (!_isEmpty(appointment)) {
    // If patient is on schedule,the note can be attached to their most recent appointment on the Schedule.
    // any appointment type other than default (0) (i.e. typeValue > 0) is already part of schedule
    // S2 does not have latest appointment, rather than changing it client side we will get it from sms
    const appointmentsFromSms = await getAppointments({
      includeStatuses: "true",
      appointmentId: appointmentReader.id(appointment)
    });
    const appointmentFromSms = _get(appointmentsFromSms, "appointment");
    const appointmentType = appointmentReader.type(appointmentFromSms);
    const appointmentStartsAt = appointmentReader.startsAt(appointmentFromSms);

    if (!_isEmpty(appointmentStartsAt)) {
      const appointmentDate = new Date(appointmentStartsAt);
      const now = new Date();
      const diffInDays = differenceInCalendarDays(now, appointmentDate);

      // If an old appointment gets attached compared to today's date then, show the
      // modal to just to make the user aware as to what has happened.
      if (diffInDays && isNewNote) {
        dispatch(noteActions.setIsNewNote(false));
        dispatch(modalsActions.showDateOfServiceConfirmation());
        return;
      }
    }

    if (appointmentType !== AppointmentTypes.DEFAULT) {
      return;
    }
  }

  // Else, if the patient is on the List but not the Schedule, the note can be attached to that
  // since Epic do not have patient list, so return
  if (isEpicEMR(emrType)) {
    return;
  }

  // check if patient is in patient list
  getPatientList().then((patients: Patients) => {
    const isPatientInList = _find(patients, (patient) => patientReader.id(patient) === patientId);

    if (!_isEmpty(isPatientInList)) {
      return;
    }

    //  Else, if (the patient is on neither the List nor the Schedule
    // â€“ show popup to add to list or schedule (non-emr only))
    dispatch(modalsActions.showAddToScheduleOrList());
  });
}

export { handleAddPatientToScheduleOrList };
