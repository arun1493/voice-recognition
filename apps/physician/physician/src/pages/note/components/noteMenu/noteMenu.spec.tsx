import _noop from "lodash/noop";
import "@testing-library/jest-dom";
import "@testing-library/jest-dom/extend-expect";
import { cleanup, RenderResult } from "@testing-library/react";

import { NoteMenu } from "./NoteMenu";
import { configureI18n } from "../../../../helpers";
import { componentSetupWithStore } from "../../../../hocs";
import { shouldMatchSnapshot } from "../../../../helpers/snapshot.spec";

const props = {
  isReadOnlyMode: true,
  onClose: _noop
};

const setup = componentSetupWithStore(() => <NoteMenu {...props} />);

describe("NoteMenu", () => {
  let component: RenderResult;

  beforeEach(() => {
    configureI18n();
    component = setup();
  });

  afterEach(() => {
    cleanup();
  });

  it("should render successfully", () => {
    const { baseElement } = component;
    expect(baseElement).toBeTruthy();
  });

  it("should contain Print'", () => {
    const { baseElement } = component;
    const element = baseElement.querySelector("[data-cy='note-menu-item-0']");
    expect(element?.innerHTML).toContain("Print");
  });

  it("should contain Delete'", () => {
    const { baseElement } = component;
    const element = baseElement.querySelector("[data-cy='note-menu-item-1']");
    expect(element?.innerHTML).toContain("Delete");
  });

  it("should match snapshot", () => {
    shouldMatchSnapshot(NoteMenu, props);
  });
});
