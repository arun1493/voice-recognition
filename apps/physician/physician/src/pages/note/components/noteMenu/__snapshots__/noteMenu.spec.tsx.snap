// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteMenu should match snapshot 1`] = `
<div
  className="dropDownMenustyled__StyledDropDownMenu-tw503w-0 gTvYNk"
  data-cy="drop-down-menu-list-item"
  onClick={[Function]}
>
  <div
    className="noteMenustyled__StyledNoteMenuContainer-sc-17cvm2-0 jnjZa-D"
    data-cy="note-menu"
  >
    <a
      className="noteMenustyled__StyledLink-sc-17cvm2-1 bSgFIj"
      data-cy="note-menu-item-0"
      href="/home"
      isDisabled={false}
      onClick={[Function]}
    >
      <div
        className="noteMenuListItemstyled__StyledNoteMenuListItems-sc-10lu9qc-0 EzATm"
      >
        <div
          className="noteMenuListItemstyled__StyledNoteMenuIcon-sc-10lu9qc-1 euPyIa"
        >
          <svg
            fill="none"
            height={24}
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clipRule="evenodd"
              d="M8 6H16V7H8V6ZM6 7V6C6 4.89543 6.89543 4 8 4H16C17.1046 4 18 4.89543 18 6V7H20C21.1046 7 22 7.89543 22 9V15C22 16.1046 21.1046 17 20 17H18V20H6V17H4C2.89543 17 2 16.1046 2 15V9C2 7.89543 2.89543 7 4 7H6ZM16 9H8H4V15H6V11H18V15H20V9H16ZM16 13V15V17V18H8V17V15V13H16ZM9 14H15V15H9V14ZM15 16H9V17H15V16Z"
              fill="#2e5f62"
              fillRule="evenodd"
              opacity="1"
            />
          </svg>
        </div>
        <p
          className="textstyled__StyledText-sc-1vedkxi-0 WgCWq"
          color="green"
          cursor="pointer"
          data-cy="note-menu-item-0"
          onClick={[Function]}
          size="sm"
        >
          Print
        </p>
      </div>
    </a>
    <a
      className="noteMenustyled__StyledLink-sc-17cvm2-1 bROybI"
      data-cy="note-menu-item-1"
      href="/home"
      isDisabled={true}
      onClick={[Function]}
    >
      <div
        className="noteMenuListItemstyled__StyledNoteMenuListItems-sc-10lu9qc-0 EzATm"
      >
        <div
          className="noteMenuListItemstyled__StyledNoteMenuIcon-sc-10lu9qc-1 euPyIa"
        >
          <svg
            fill="none"
            height={24}
            viewBox="0 0 24 24"
            width={24}
          >
            <path
              clipRule="evenodd"
              d="M8.15558 4.18772C8.47659 3.46547 9.19283 3 9.98321 3H14.0168C14.8072 3 15.5234 3.46547 15.8444 4.18772L16.6499 6H16.9913H19C19.5523 6 20 6.44772 20 7C20 7.55228 19.5523 8 19 8H18.8133L17.7693 19.1859C17.6734 20.214 16.8105 21 15.778 21H8.22202C7.18946 21 6.32664 20.214 6.23068 19.1859L5.18667 8H5C4.44772 8 4 7.55228 4 7C4 6.44772 4.44772 6 5 6H7.00869H7.35013L8.15558 4.18772ZM16 8H16.8046L15.778 19H8.22202L7.19536 8H8H16ZM14.0168 5L14.4612 6H9.53876L9.98321 5H14.0168ZM9.49852 9.46127C9.47782 9.18604 9.23784 8.98144 8.96251 9.0043C8.68719 9.02716 8.48077 9.26881 8.50148 9.54404L9.10148 17.5206C9.12218 17.7958 9.36216 18.0004 9.63749 17.9775C9.91281 17.9547 10.1192 17.713 10.0985 17.4378L9.49852 9.46127ZM15.0375 9.0043C14.7622 8.98144 14.5222 9.18604 14.5015 9.46127L13.9015 17.4378C13.8808 17.713 14.0872 17.9547 14.3625 17.9775C14.6378 18.0004 14.8778 17.7958 14.8985 17.5206L15.4985 9.54404C15.5192 9.26881 15.3128 9.02716 15.0375 9.0043ZM12.5 9.50266C12.5 9.22651 12.2761 9.00266 12 9.00266C11.7239 9.00266 11.5 9.22651 11.5 9.50266V17.492C11.5 17.7682 11.7239 17.992 12 17.992C12.2761 17.992 12.5 17.7682 12.5 17.492V9.50266Z"
              fill="#2e5f62"
              fillRule="evenodd"
              opacity="1"
            />
          </svg>
        </div>
        <p
          className="textstyled__StyledText-sc-1vedkxi-0 WgCWq"
          color="green"
          cursor="pointer"
          data-cy="note-menu-item-1"
          onClick={[Function]}
          size="sm"
        >
          Delete
        </p>
      </div>
    </a>
  </div>
</div>
`;
